lntest/node.go

```diff
@@ -0,0 +1,679 @@
+// designated for the current lightning network test. If there haven't been any
+// test instances created, the default ports are used. Otherwise, in order to
+// support multiple test nodes running at once, the p2p and rpc port are
+// incremented after each initialization.
+func generateListeningPorts() (int, int) {
+	var p2p, rpc int
+	if numActiveNodes == 0 {
+		p2p = defaultNodePort
+		rpc = defaultClientPort
+	} else {
+		p2p = defaultNodePort + (2 * numActiveNodes)
+		rpc = defaultClientPort + (2 * numActiveNodes)
+	}
+
+	return p2p, rpc
+}
+
+type nodeConfig struct {
+	RPCConfig *rpcclient.ConnConfig
+	NetParams *chaincfg.Params
+	BaseDir   string
+	ExtraArgs []string
+
+	DataDir      string
+	LogDir       string
+	TLSCertPath  string
+	TLSKeyPath   string
+	AdminMacPath string
+	ReadMacPath  string
+	P2PPort      int
+	RPCPort      int
+}
+
+func (cfg nodeConfig) P2PAddr() string {
+	return net.JoinHostPort("127.0.0.1", strconv.Itoa(cfg.P2PPort))
+}
+
+func (cfg nodeConfig) RPCAddr() string {
+	return net.JoinHostPort("127.0.0.1", strconv.Itoa(cfg.RPCPort))
+}
+
+func (cfg nodeConfig) DBPath() string {
+	return filepath.Join(cfg.DataDir, cfg.NetParams.Name, "bitcoin/channel.db")
+}
+
+// genArgs generates a slice of command line arguments from the lightning node
+// config struct.
+func (cfg nodeConfig) genArgs() []string {
+	var args []string
+
+	switch cfg.NetParams {
+	case &chaincfg.TestNet3Params:
+		args = append(args, "--bitcoin.testnet")
+	case &chaincfg.SimNetParams:
+		args = append(args, "--bitcoin.simnet")
+	case &chaincfg.RegressionNetParams:
+		args = append(args, "--bitcoin.regtest")
+	}
+
+	encodedCert := hex.EncodeToString(cfg.RPCConfig.Certificates)
+	args = append(args, "--bitcoin.active")
+	args = append(args, "--nobootstrap")
+	args = append(args, "--noencryptwallet")
+	args = append(args, "--debuglevel=debug")
+	args = append(args, fmt.Sprintf("--bitcoin.rpchost=%v", cfg.RPCConfig.Host))
+	args = append(args, fmt.Sprintf("--bitcoin.rpcuser=%v", cfg.RPCConfig.User))
+	args = append(args, fmt.Sprintf("--bitcoin.rpcpass=%v", cfg.RPCConfig.Pass))
+	args = append(args, fmt.Sprintf("--bitcoin.rawrpccert=%v", encodedCert))
+	args = append(args, fmt.Sprintf("--rpcport=%v", cfg.RPCPort))
+	args = append(args, fmt.Sprintf("--peerport=%v", cfg.P2PPort))
+	args = append(args, fmt.Sprintf("--logdir=%v", cfg.LogDir))
+	args = append(args, fmt.Sprintf("--datadir=%v", cfg.DataDir))
+	args = append(args, fmt.Sprintf("--tlscertpath=%v", cfg.TLSCertPath))
+	args = append(args, fmt.Sprintf("--tlskeypath=%v", cfg.TLSKeyPath))
+	args = append(args, fmt.Sprintf("--configfile=%v", cfg.DataDir))
+	args = append(args, fmt.Sprintf("--adminmacaroonpath=%v", cfg.AdminMacPath))
+	args = append(args, fmt.Sprintf("--readonlymacaroonpath=%v", cfg.ReadMacPath))
+	args = append(args, fmt.Sprintf("--externalip=%s", cfg.P2PAddr()))
+	args = append(args, fmt.Sprintf("--trickledelay=%v", trickleDelay))
+
```

Comment:
Note that in master, we now manually set `defaultnumconfs`.

Reply:
Rebased.